node/src/components/linear_chain.rs

```diff
@@ -444,22 +461,32 @@ where
                     self.signature_cache.insert(*signatures.clone());
                 }
                 // Check if the validator is bonded in the era in which the block was created.
-                effect_builder
-                    .is_bonded_validator(fs.era_id, fs.public_key)
-                    .map(|is_bonded| {
-                        if is_bonded {
-                            Ok((maybe_signatures, fs, is_bonded))
-                        } else {
-                            Err((maybe_signatures, fs))
-                        }
-                    })
+                // TODO: Use protocol version that is valid for the block's height.
+                let protocol_version = self.protocol_version;
+                let initial_state_root_hash = self.initial_state_root_hash;
+                let activation_era_id = self.activation_era_id;
+                async move {
+                    match effect_builder
+                        .is_bonded_validator(
+                            &fs.public_key,
+                            fs.era_id,
+                            activation_era_id,
+                            initial_state_root_hash,
+                            protocol_version,
+                        )
+                        .await
```

Comment:
I wonder whether the whole `is_bonded_in_future_era` logic should be moved in there, too?

Reply:
I agree, it should be unified.

Reply:
Also, since this logic is in `effect.rs` now, can we use it in the era supervisor as well, and deduplicate it? (Probably not in this PR, though.)

Reply:
Could be quite complex. I had a brief look at the logic in the Era Supervisor, but it seems that it's a bit different there. Deduplication would be great, though.

Reply:
I actually did both things :+1: